influence_benchmark/environment_vectorized/trajectory_queue.py

```diff
@@ -164,87 +188,65 @@ def populate(self, iter_step: int, eval: bool = False, allow_id_to_see_tool_call
+                if "possible_env_vars" in self.main_config:
+                    possible_vars = self.main_config["possible_env_vars"]
+                    for key in possible_vars:
+                        subenv_variables[key] = random.choice(possible_vars[key])
 
                 # Each subenv has n_trajs_to_sample_per_subenv trajectories which have to be generated with the same initial state
                 for traj_id in range(n_trajs_to_sample_per_subenv):
-                    subenv = gen_subenv_from_configs(subenv_args, subenv_id, subenv_config)
+                    subenv = self.gen_subenv_from_configs(env_name, subenv_id, env_config, subenv_variables)
                     subenv["traj_id"] = traj_id
                     self.put(subenv)
+
         # Each process should have enough end-of-queue signals. If the queue is empty and we call get(), it will stall
-        for _ in range(len(self.devices) * self.env_args["num_envs_per_device"]):
+        for _ in range(len(self.devices) * self.num_envs_per_device):
             self.put("END_OF_QUEUE")
 
+    def gen_subenv_from_configs(self, env_name: str, subenv_id: str, env_config: dict, subenv_variables: dict):
+        """
+        Generate a subenvironment from the given environment and subenvironment configurations.
+        """
+        subenv_dict = {}
+        formatting_vars = {**subenv_variables, **env_config}
+
+        initial_state_history = []
+        for message in env_config["histories"][subenv_id]:
+            # NOTE: If this assert isn't triggered, we can remove variables from argument to this class and State class, and clean
+            # up stuff in trajectory_queue.py
+            assert count_format_fields(message["content"]) == 0
+            formatted_message = message["content"].format_map(formatting_vars).strip()
+            initial_state_history.append({"role": message["role"], "content": formatted_message})
+
+        subenv_dict["environment"] = Environment(
+            history_id=subenv_id,
+            state_config=self.main_config["state_config"],
+            initial_messages=initial_state_history,
+            format_vars=formatting_vars,
+            max_turns=self.max_turns,
+            env_name=env_name,
+        )
+
+        # Set up AssessorModels and Character
+        for key, system_prompt in self.system_prompts.items():
+            config = self.main_config[key + "_config"]
+            if "preference" in key or "influence" in key or "character" in key:
+                assert count_format_fields(system_prompt) != 0, "Maybe this is a deepcopy bug?"
+
+            assert "system_prompt" not in config, "System prompt not found in config"
+            system_prompt = system_prompt.format_map(formatting_vars).strip()
+
+            class_name = Character if "character" in key else AssessorModel
+
+            # TODO: do we want to allow_id_to_see_tool_calls for all AssessorModels? Probably not? This is how it was done in the original code AFAIK
```

Comment:
Just wanted to flag this comment, seems quite important – should all the assessor models see the tool calls, or just the preference model?